### PR TITLE
Chapter 2 4

### PR DESCRIPTION
**Pull request by DaddyWesker**
Okay, I've finished chapter 2.4

### CODE REVIEW

File: SICP_book/chapter_2_4.metta [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+    (sqrt! $x))
+
+(= (atan $x $y)
+    (atan! $x $y))
+
+(= (cos $x)
+    (cos! $x))
+
+(= (sin $x)
+    (sin! $x))
```
Comment (vsbogd) [inline SICP_book/chapter_2_4.metta:26]: Not sure why do you need to duplicate these functions.

File: SICP_book/chapter_2_4.metta [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+
```
Comment (vsbogd) [inline SICP_book/chapter_2_4.metta:27]: Not sure why do you need to duplicate these functions.
```suggestion

(= (sqrt $x)
    (sqrt! $x))

(= (atan $x $y)
    (atan! $x $y))

(= (cos $x)
    (cos! $x))

(= (sin $x)
    (sin! $x))

```

File: SICP_book/additional_funcs.py [hunk 3]
```diff
@@ -18,11 +20,26 @@ def getCurTime():
...
 @register_atoms
 def my_glob_atoms():
     return {
         'randomint!': OperationAtom("randomint!", getRandInt),
         'timems!': OperationAtom("timems!", getCurTime),
         '//': OperationAtom("//", quotient),
+        'sqrt!': OperationAtom("sqrt!", getSqrt),
+        'atan!': OperationAtom("atan!", getAtan),
+        'cos!': OperationAtom("cos!", getCos),
+        'sin!': OperationAtom("sin!", getSin),
```
[the remark this answers is not in the document]
Comment (DaddyWesker) [reply SICP_book/additional_funcs.py:44]: What you mean by side effects in this case?

**Review by vsbogd — CHANGES_REQUESTED** [review]

File: SICP_book/additional_funcs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vsbogd) [reply SICP_book/additional_funcs.py:44]: I mean code of these functions doesn't modify the program state outside their local scope. Why did you used exclamation mark in names? What is the convention you use?

[the remark this answers is not in the document]
Comment (DaddyWesker) [reply SICP_book/additional_funcs.py:44]: Well I just thought it is some sort of rule to use ! symbol for imported function.

[the remark this answers is not in the document]
Comment (DaddyWesker) [reply SICP_book/additional_funcs.py]: Anyway, done.

File: SICP_book/chapter_2_4.metta (hunk 2 above)
Comment (DaddyWesker) [reply SICP_book/chapter_2_4.metta]: Done.

File: SICP_book/chapter_2_4.metta [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+;    (apply-generic_message magnitude &m_z2)
+;    3)
+
+;!(assertEqual
+;    (apply-generic_message angle &m_z1)
+;    0.3805063771123649)
+
+;!(assertEqual
+;    (apply-generic_message angle &m_z2)
+;    4)
```
Comment (vsbogd) [inline SICP_book/chapter_2_4.metta:707]: Not sure why some code is commented out? Does it mean asserts are not stable or it is a performance issue?

Comment (DaddyWesker) [reply SICP_book/chapter_2_4.metta]: My bad. They work fine. I've pushed commit without ';'

**Review by vsbogd — APPROVED** [review]

**Pull request merged**